src/frontend/src/scheduler/worker_node_manager.rs

```diff
@@ -183,6 +186,55 @@ impl WorkerNodeManager {
             .remove(fragment_id)
             .unwrap();
     }
+
+    /// Returns vnode mapping for serving.
+    /// If there are available serving workers, use them.
+    /// Otherwise fall back to streaming workers, even they have set `disable_serving`.
+    /// `fragment_id` is a hint providing streaming vnode mapping.
+    pub fn serving_vnode_mapping(
```

Comment:
The behavior of falling back to streaming vnodes is weird to me. The goal of separating serving nodes and streaming nodes is for workload isolation, and blindly falling back to streaming nodes may introduce instability and misuse, so I would prefer disabling falling back first, and add some config to enable falling back to streaming when necessary.